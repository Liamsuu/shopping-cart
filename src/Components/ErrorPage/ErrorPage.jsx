import { Link } from "react-router-dom";

export default function ErrorPage() {
  return (
    <div>
      <p>
        This path does not exist, here are some useful pages: <br />
        <Link to="/">Homepage</Link>
        <br />
        <Link to="cart">Cart</Link>
      </p>
    </div>
  );
}
