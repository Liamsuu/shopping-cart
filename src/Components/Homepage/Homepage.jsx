import NavBar from "../NavBar/NavBar";

function Homepage() {
  return (
    <>
      <NavBar />
    </>
  );
}

export default Homepage;
