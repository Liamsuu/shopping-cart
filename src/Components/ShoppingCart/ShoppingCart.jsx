import styled from "styled-components";
import NavBar from "../NavBar/NavBar";
import BasketProduct from "../Product/BasketProduct";
import { useLocation } from "react-router-dom";

const BasketListWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 5rem;
`;

const BasketList = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: #f9fafb;
  width: 40%;
  padding: 0rem 1rem 2rem 1rem;
  border-radius: 5px;
`;

export default function ShoppingCart() {
  const location = useLocation();
  const itemsInCart = location.state.itemsInCart;
  const numItemsInCart = location.state.numItemsInCart;

  return (
    <>
      <NavBar itemsInCart={itemsInCart} numItemsInCart={numItemsInCart} />
      <BasketListWrapper>
        <BasketList>
          <h2 style={{ marginRight: "auto" }}>Basket</h2>

          {itemsInCart.map((productObj) => {
            return (
              <BasketProduct
                key={productObj.key}
                productImg={productObj.image}
                productName={productObj.title}
                productPrice={productObj.price}
              />
            );
          })}
        </BasketList>
      </BasketListWrapper>
    </>
  );
}
