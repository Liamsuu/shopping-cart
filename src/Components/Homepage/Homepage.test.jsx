import { describe, it, expect } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { RouterProvider, createBrowserRouter } from "react-router-dom";
import Homepage from "./Homepage";
import ErrorPage from "../ErrorPage/ErrorPage";
import ShoppingCart from "../ShoppingCart/ShoppingCart";

const router = createBrowserRouter([
  {
    path: "/",
    element: <Homepage />,
    errorElement: <ErrorPage />,
  },
  {
    path: "cart",
    element: <ShoppingCart />,
  },
]);

describe("Homepage component", () => {
  it("Renders homepage", () => {
    const { container } = render(<RouterProvider router={router} />);
    expect(container).toMatchSnapshot();
  });

  it("Increments basket quantity by one on click", async () => {
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();
    await waitFor(() =>
      expect(screen.getByText("Stay Unique")).toBeInTheDocument()
    );
    const addToCartBtn = screen.getAllByRole("button", { name: "Add to cart" });
    await user.click(addToCartBtn[0]);
    expect(screen.getByText("1")).toBeInTheDocument();
  });
  it("Increments basket quantity when same product is clicked multiple times", async () => {
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();
    await waitFor(() =>
      expect(screen.getByText("Stay Unique")).toBeInTheDocument()
    );
    const addToCartBtn = screen.getAllByRole("button", { name: "Add to cart" });
    await user.click(addToCartBtn[0]);
    await user.click(addToCartBtn[0]);
    expect(screen.getByText("2")).toBeInTheDocument();
  });
  it("Increments basket quantity when different product are added to basket", async () => {
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();
    await waitFor(() =>
      expect(screen.getByText("Stay Unique")).toBeInTheDocument()
    );
    const addToCartBtn = screen.getAllByRole("button", { name: "Add to cart" });
    await user.click(addToCartBtn[0]);
    await user.click(addToCartBtn[1]);
    await user.click(addToCartBtn[2]);
    await user.click(addToCartBtn[3]);
    await user.click(addToCartBtn[4]);
    await user.click(addToCartBtn[5]);
    expect(screen.getByText("6")).toBeInTheDocument();
  });
  it("Redirects to new page when basket is clicked", async () => {
    render(<RouterProvider router={router} />);
    const user = userEvent.setup();
    await waitFor(() =>
      expect(screen.getByText("Stay Unique")).toBeInTheDocument()
    );
    const basketButton = screen.getByRole("link", { name: "cart" });
    user.click(basketButton);
    await waitFor(() => expect(screen.getByText("Basket")).toBeInTheDocument());
  });
});
