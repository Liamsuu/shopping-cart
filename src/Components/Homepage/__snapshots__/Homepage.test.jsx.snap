// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Homepage component > Renders homepage 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
